docker: optionally package with debug symbols

rcoup opened the pull request:
Enable debug symbols in "ubuntu-full" docker images for Proj and GDAL non-release builds. This allows easier issue reproduction against master.

* debug symbols are in external files to minimise process memory
* symbols included for proj & gdal
* enable via `--with-debug-symbols/--without-debug-symbols` argument to build scripts
* included by default for non-release builds
* adds about 200MB to the image size

**Note:** Currently only applies to ubuntu-full, if there's interest could expand it to other image flavours.

I also simplified the ubuntu-full Dockerfile by moving proj & gdal builds to scripts, and added some more command safety: mostly driven by shellcheck and changing `;` → `&&` for consecutive commands.

Todo:

- [ ] decide whether to generalise the shell build scripts for the other image flavours.
- [ ] decide whether debug symbols would be useful for any other image flavours
- [ ] test building on not-my-machine

rouault commented:
Thanks for that!

From what I can see for GDAL builds, the effect of setting CFLAGS=-ggdb disables -O2, which is probably not what we want since the binaries will be sluggish. By default, CFLAGS for GDAL are already "-g -O2". So there is probably no need to override them for GDAL, or use "-ggdb -O2".
By the way I've never used -ggdb. Not sure what it brings compared to plain -g.

rouault commented:
Got an error
```
Downloaded: 335 files, 498M in 7m 10s (1.16 MB/s)
The command '/bin/sh -c mkdir -p ${PROJ_INSTALL_PREFIX}/share/proj     && wget --no-verbose --mirror https://cdn.proj.org/     && rm -f cdn.proj.org/*.js     && rm -f cdn.proj.org/*.css     && mv cdn.proj.org/* ${PROJ_INSTALL_PREFIX}/share/proj/     && rmdir cdn.proj.org' returned a non-zero code: 8
```

rouault commented:
> Got an error
> 
> ```
> 
> ```

This is likely due to some failed attempts at fetching some unneeded files:
```
https://cdn.proj.org/robots.txt:
2020-06-17 15:55:15 ERROR 404: Not Found.
2020-06-17 15:55:20 URL:https://cdn.proj.org/geotiff.bundle.min.js [223236/223236] -> "cdn.proj.org/geotiff.bundle.min.js" [1]
https://cdn.proj.org/'%20+%20props.name%20+%20':
2020-06-17 15:55:21 ERROR 404: Not Found.
```
I think I hit that issue and decided to ignore the return code for that reason.
Perhaps a better way would be to use the projsync utility that comes with PROJ now with something like "projsync --system-directory -all"

rcoup commented:
> Perhaps a better way would be to use the projsync utility that comes with PROJ now with something like "projsync --system-directory -all"

Done

> From what I can see for GDAL builds, the effect of setting CFLAGS=-ggdb disables -O2, which is probably not what we want since the binaries will be sluggish. By default, CFLAGS for GDAL are already "-g -O2". So there is probably no need to override them for GDAL, or use "-ggdb -O2".

Nice catch, I'd been fiddling with Proj where that wasn't the case. Fixed

> By the way I've never used -ggdb. Not sure what it brings compared to plain -g.

Probably nothing for Linux, I swapped to `-g`.

I suspect Proj installing with `PREFIX=/` was an oversight, so I fixed that.

rouault commented:
By the way, what does bh- means in bh-gdal.sh ?

rcoup commented:
ah, build helper. It was `build-gdal.sh`, then I realised that having `build.sh` and `build-gdal.sh` and `build-proj.sh` beside each other wasn't helpful to figure out which to run 😄

rcoup commented:
As in the Todo, if you think it's useful I could try and spend some time normalising the other builds to all use a single set of helper scripts.

rouault commented:
> spend some time normalising the other builds to all use a single set of helper scripts.

I don't know how much it is feasible. There might be subtle differences between the builds

rcoup commented:
Sure. If someone's enthusiastic I guess they can pick up the todo's 😄
